src/RA_StringUtils.h

```diff
@@ -25,65 +25,486 @@ _NODISCARD std::string Narrow(_In_ const std::string& wstr);
+                AppendPrintf(++pScan, std::forward<Ts>(args)...);
+                break;
         }
+    }
+
+    std::string ToString() const
+    {
+        std::string sResult;
+        AppendToString(sResult);
+        return sResult;
+    }
 
-        sFormatted.resize(ra::to_unsigned(nNeeded));
+    void AppendToString(_Inout_ std::string& sResult) const;
+
+    std::wstring ToWString() const
+    {
+        std::wstring sResult;
+        AppendToWString(sResult);
+        return sResult;
     }
 
-    return sFormatted;
+    void AppendToWString(_Inout_ std::wstring& sResult) const;
+
+private:
+    bool m_bPrepareWide;
+
+    struct PendingString
+    {
+        union
+        {
+            const std::string* String;
+            const std::wstring* WString;
+
+            struct
+            {
+                const char* Pointer;
+                size_t Length;
+            } Char;
+
+            struct
+            {
+                const wchar_t* Pointer;
+                size_t Length;
+            } WChar;
+        } Ref{};
+
+        std::string String;
+        std::wstring WString;
+
+        enum class Type
+        {
+            String,
+            WString,
+            StringRef,
+            WStringRef,
+            CharRef,
+            WCharRef,
+        };
+        Type DataType{ Type::String };
+    };
+
+    mutable std::vector<PendingString> m_vPending;
+};
+
+template<typename... Ts>
+void AppendString(std::string& sString, Ts... args)
+{
+    StringBuilder builder;
+    builder.Append(std::forward<Ts>(args)...);
+    builder.AppendToString(sString);
 }
-// This one does depend on the character set
-template<typename Arithmetic, class = std::enable_if_t<std::is_arithmetic_v<Arithmetic>>>
-_NODISCARD inline auto to_tstring(_In_ Arithmetic a) noexcept
+
+template<typename... Ts>
+void AppendWString(std::wstring& sString, Ts... args)
 {
-#if _MBCS
-    return std::to_string(a);
-#elif _UNICODE
-    return std::to_wstring(a);
-#else
-#error Unknown character set detected, only MultiByte and Unicode character sets are supported!
-#endif // _MBCS
-} // end function to_tstring
+    StringBuilder builder(true);
+    builder.Append(std::forward<Ts>(args)...);
+    builder.AppendToWString(sString);
+}
+
+template<typename... Ts>
+std::string BuildString(Ts... args)
+{
+    StringBuilder builder;
+    builder.Append(std::forward<Ts>(args)...);
+    return builder.ToString();
+}
+
+template<typename... Ts>
+std::wstring BuildWString(Ts... args)
+{
+    StringBuilder builder(true);
+    builder.Append(std::forward<Ts>(args)...);
+    return builder.ToWString();
+}
+
+/// <summary>
+/// Constructs a <see cref="std::basic_string" /> from a format string and parameters
+/// </summary>
+template<typename CharT, typename = std::enable_if_t<is_char_v<CharT>>, typename... Ts>
+_NODISCARD inline auto StringPrintf(_In_z_ _Printf_format_string_ const CharT* const __restrict sFormat, Ts... args)
```

Comment:
`args` should be passed by `rvalue reference`.

Reply:
Doesn't work. This is the entry point function and most of the parameter are passed ad hoc.
```
cannot convert from 'const size_t' to 'const size_t &&'
cannot convert from 'const int' to 'const int &&'
cannot convert from 'const char*' to 'const char* const &&'
...
```

Reply:
Makes sense. Rvalue refs cant be const. In the case of c strings, it cant be a const pointer (const after *)